### PR TITLE
Gateway-related conformance tests

### LINKED ISSUE #2777: all conformance tests for `v1beta1.Gateway` are run and required for every PR
### Problem statement

With the Gateway API `v1beta1` a new set of conformance tests related to the gateway has been introduced. We should support all those tests to be fully conformant with the API.

### Tests to be conformant with:

- [x] GatewaySecretInvalidReferenceGrant
- [x] GatewaySecretMissingReferenceGrant
- [x] GatewaySecretReferenceGrantAllInNamespace
- [x] GatewaySecretReferenceGrantSpecific

### Acceptance criteria

- [ ] All the above tests are passing

### Additional Notes

This issue is blocked-by #2775

### PR DESCRIPTION
**Pull request by mlavacca**
**What this PR does / why we need it**:
All the conformance tests related to the gateways are passing and enabled by default.

**Which issue this PR fixes**:
Fixes #2777 

**Special notes for your reviewer**:

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [x] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR

### CODE REVIEW

**Comment by mlavacca** [comment]
**Note** for the reviewers: this PR only aims at implementing the minimal set of specifics to pass the conformance tests. Some corner cases in the gateway status update are not handled here and are out of scope for this PR. I created two different issues to cover them:
* #2785 
* #2786

File: internal/controllers/gateway/gateway_utils.go [hunk 1]
```diff
@@ -342,23 +343,76 @@ func getListenerStatus(
...
+			resolvedRefReason = string(gatewayv1alpha2.ListenerReasonResolvedRefs)
+			for _, certRef := range listener.TLS.CertificateRefs {
+				// only secrets are supported as certificate references
+				if (certRef.Group != nil && (*certRef.Group != "core" && *certRef.Group != "")) ||
+					(certRef.Kind != nil && *certRef.Kind != "Secret") {
+					resolvedRefReason = string(gatewayv1alpha2.ListenerReasonInvalidCertificateRef)
+					break
+				}
+				// if the certificate is in the same namespace of the gateway, no ReferenceGrant is needed
+				if certRef.Namespace == nil {
```
Comment (shaneutt) [inline internal/controllers/gateway/gateway_utils.go:374]: Technically certificate refs could have a ref with a namespace that is the same as the `Gateway` but that namespace explicitly called out in the listener, right? Even if we don't expect this with our current controller, it would be valid behavior in the future so we might as well add the conditional here:

```suggestion
				if certRef.Namespace == nil || *certRef.Namespace == gatewayv1alpha2.Namespace(gateway.Namespace) {
```

Is this right or am I missing something? :thinking:

File: internal/controllers/gateway/gateway_utils.go [hunk 2]
```diff
@@ -342,23 +343,76 @@ func getListenerStatus(
...
+					LastTransitionTime: metav1.Now(),
+					Reason:             resolvedRefReason,
+				},
+				metav1.Condition{
+					Type:               string(gatewayv1alpha2.ListenerConditionReady),
+					Status:             conditionStatus,
+					ObservedGeneration: gateway.Generation,
+					LastTransitionTime: metav1.Now(),
+					Reason:             string(gatewayv1alpha2.ListenerReasonReady),
+					Message:            "the listener is not ready and cannot route requests",
```
Comment (shaneutt) [inline internal/controllers/gateway/gateway_utils.go:410]: Unless the `resolvedRefReason` comes back with something other than `ListenerReasonResolvedRefs` this condition could be `metav1.ConditionTrue`, in which case this message doesn't make sense right? Should we not have a permutation of this message for the `ConditionTrue` case?

File: internal/controllers/gateway/gateway_utils.go [hunk 3]
```diff
@@ -398,6 +454,42 @@ func getListenerStatus(
...
+					// the granted resource name, then return a reason "ResolvedRefs"
+					if to.Name == nil || string(*to.Name) == string(certRef.Name) {
+						return string(gatewayv1alpha2.ListenerReasonResolvedRefs)
+					}
+				}
+			}
+		}
+	}
+	// if no grants have been found for the reference, return an "InvalidCertificateRef" reason
+	return string(gatewayv1alpha2.ListenerReasonInvalidCertificateRef)
```
Comment (shaneutt) [inline internal/controllers/gateway/gateway_utils.go:490]: It doesn't seem that we added any unit tests for this function, I recommend we add some.

**Review by shaneutt — CHANGES_REQUESTED** [review]
Looking good!

Mostly minor comments and questions and a request for some more testing.

File: internal/controllers/gateway/gateway_utils.go (hunk 1 above)
Comment (mlavacca) [reply internal/controllers/gateway/gateway_utils.go:374]: yes, you are right! I'm going to add this condition

**Review by shaneutt — APPROVED** [review]
